### PR TITLE
[feature](partial update) add failure tolerance for strict mode partial update stream load

### PR DESCRIPTION
**Pull request by bobhan1**
## Proposed changes

1. add the number of newly inserted rows to num_filtered_rows in strict mode partial update stream load
2. use delete_bitmap to delete the newly inserted rows in strict mode partial update
3. If num_rows_load_filtered/num_selected_rows > max_filter_ratio, the strict mode partial update stream load will fail.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by bobhan1** [comment]
run buildall

File: be/src/olap/rowset/segment_v2/segment_writer.cpp [hunk 1]
```diff
@@ -408,6 +417,25 @@ Status SegmentWriter::append_block_with_partial_content(const vectorized::Block*
...
+
+    if (_tablet_schema->is_strict_mode() && num_rows_filtered > 0) {
+        // size_t dest_size = full_block.columns();
+        // full_block.insert(vectorized::ColumnWithTypeAndName(
+        //         std::move(filter_column), std::make_shared<vectorized::DataTypeUInt8>(),
+        //         "filter column"));
+        // RETURN_IF_ERROR(vectorized::Block::filter_block(&full_block, dest_size, dest_size));
+        auto& delete_column_data =
+                assert_cast<vectorized::ColumnInt8*>(
+                        full_block.get_by_name(DELETE_SIGN).column->assume_mutable().get())
```
Comment (zhannngchen) [inline be/src/olap/rowset/segment_v2/segment_writer.cpp:429]: use delete bitmap instead

**Comment by bobhan1** [comment]
run buildall

File: be/src/olap/rowset/segment_v2/segment_writer.cpp [hunk 2]
```diff
@@ -364,6 +365,11 @@ Status SegmentWriter::append_block_with_partial_content(const vectorized::Block*
     }
     std::vector<std::unique_ptr<SegmentCacheHandle>> segment_caches(specified_rowsets.size());
     // locate rows in base data
+
+    int64_t num_rows_filtered = 0;
+
+    std::vector<size_t> delete_idx;
```
[the remark this answers is not in the document]
Comment (bobhan1) [reply be/src/olap/rowset/segment_v2/segment_writer.cpp:371]: I forget to delete this.

**Review by zhannngchen — DISMISSED** [review]
LGTM

**Review by dataroaring — DISMISSED** [review]
LGTM

File: be/src/olap/rowset/beta_rowset_writer.h [hunk 3]
```diff
@@ -240,6 +243,8 @@ class BetaRowsetWriter : public RowsetWriter {
 
     int64_t _delete_bitmap_ns = 0;
     int64_t _segment_writer_ns = 0;
+
+    DeltaWriter* _delta_writer = nullptr;
```
Comment (kaijchen) [inline be/src/olap/rowset/beta_rowset_writer.h:247]: We should not expose delta writer in rowset writer.

File: be/src/olap/rowset/beta_rowset_writer.cpp [hunk 4]
```diff
@@ -890,6 +890,10 @@ Status BetaRowsetWriter::_flush_segment_writer(std::unique_ptr<segment_v2::Segme
     segstat.index_size = index_size + (*writer)->get_inverted_index_file_size();
     segstat.key_bounds = key_bounds;
 
+    if (_delta_writer != nullptr) {
+        _delta_writer->update_filtered_rows((*writer)->num_rows_filtered());
```
Comment (kaijchen) [inline be/src/olap/rowset/beta_rowset_writer.cpp:894]: You should store this value in rowset writer and get it from delta writer instead.
i.e. Call `rowset_writer->get_num_rows_filtered()` in DeltaWriter.

**Review by kaijchen — CHANGES_REQUESTED** [review]

**Comment by bobhan1** [comment]
run buildall

File: be/src/olap/rowset/segment_v2/segment_writer.cpp [hunk 5]
```diff
@@ -408,6 +412,7 @@ Status SegmentWriter::append_block_with_partial_content(const vectorized::Block*
     auto mutable_full_columns = full_block.mutate_columns();
     RETURN_IF_ERROR(fill_missing_columns(mutable_full_columns, use_default_or_null_flag,
                                          has_default_or_nullable));
+
```
Comment (kaijchen) [inline be/src/olap/rowset/segment_v2/segment_writer.cpp:415]: Please do not change unrelated lines (to avoid merge conflict)

```suggestion
```

File: be/src/olap/delta_writer.h [hunk 6]
```diff
@@ -124,6 +124,10 @@ class DeltaWriter {
 
     int64_t total_received_rows() const { return _total_received_rows; }
 
+    int64_t DeltaWriter::num_rows_filtered() {
```
Comment (kaijchen) [inline be/src/olap/delta_writer.h:127]: This does not compile, please move it to `.cpp` file.

In header, it should just be 

```cpp
int64_t num_rows_filtered() const;
```

In cpp file (add the const)

```cpp
int64_t DeltaWriter::num_rows_filtered() const {
    return _rowset_writer == nullptr ? 0 : _rowset_writer->num_rows_filtered();
}
```

**Review by kaijchen — APPROVED** [review]
LGTM, thanks @bobhan1.

However, let's better merge this after #21635. You can run CI first.
There will be some conflicts, but since the change to rowset writer is small in this PR, it should be easy to solve.

**Comment by kaijchen** [comment]
run buildall

**Comment by kaijchen** [comment]
run buildall

**Comment by bobhan1** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 55.9 seconds
 stream load tsv:          506 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          66 seconds loaded 1101869774 Bytes, about 15 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17162431651 Bytes

**Review by liaoxin01 — APPROVED** [review]
LGTM

**Review by zhannngchen — APPROVED** [review]
LGTM
